components/candle/masteryService.ts

```diff
@@ -75,52 +75,78 @@ export class MasteryService {
         }
     }
 
-    getCompletionData(
-        locationParentId: string,
-        subLocationId: string,
-        gameVersion: GameVersion,
+    /**
+     * Get generic completion data stored in a user's profile. Called by both `getLocationCompletion` and `getFirearmCompletion`.
+     * @param userId The id of the user.
+     * @param gameVersion The game version.
+     * @param completionId An Id used to look up completion data in the user's profile. Can be `parentLocationId` or `progressionKey`.
+     * @param maxLevel The max level for this progression.
+     */
+    private getCompletionData(
         userId: string,
-    ): CompletionData {
-        if (!this.masteryData.has(locationParentId)) {
-            return undefined
-        }
-
-        //Get the mastery data
-        const masteryData: MasteryPackage =
-            this.masteryData.get(locationParentId)
-
+        gameVersion: GameVersion,
+        completionId: string,
+        maxLevel: number,
+    ) {
         //Get the user profile
         const userProfile = getUserData(userId, gameVersion)
 
-        //Gather all required data
-        const lowerCaseLocationParentId = locationParentId.toLowerCase()
-
-        userProfile.Extensions.progression.Locations[
-            lowerCaseLocationParentId
-        ] ??= {
+        // Generate default completion before trying to acquire it
+        userProfile.Extensions.progression.Locations[completionId] ??= {
             Xp: 0,
             Level: 1,
         }
 
         const locationData =
-            userProfile.Extensions.progression.Locations[
-                lowerCaseLocationParentId
-            ]
-
-        const maxLevel = masteryData.MaxLevel || 20
+            userProfile.Extensions.progression.Locations[completionId]
 
         const nextLevel: number = Math.max(
             0,
             Math.min(locationData.Level + 1, maxLevel),
         )
-        const nextLevelXp: number = xpRequiredForLevel(nextLevel)
+
+        const nextLevelXp: number = completionId.includes("LOCATION")
```

Comment:
I think this is a sketchy bit of code. Using completionId for two different things, and then making an assumption like this. I much rather see a boolean flag or an enum to determine which of the two to use. On top of that, how does that work for the lines above when completionId is not a location? Aren't sniper levels also locations?

Reply:
I did this because I did not think adding an extra function argument is necessary because it can be inferred from the `completionId` passed. I can add it if needed.

When ` getLocationCompletion` is called on sniper locations, they will fail the "has mastery" check and return `undefined`, so `getCompletionData` won't be called on them. 

If the completionId is not a location, then it is a firearm. Each sniper mission has three firearms, one for each character, and they each have their own mastery, so that's nine masteries in total. The lines above will work for them as well because all it needs is an identifier for each mastery progression, which is the `completionId`. They will be stored as objects like `"FIREARMS_SC_HERO_SNIPER_KNIGHT":{"Xp":0,"Level":1}` in `Extensions.progression.Locations`. (Even though they are not locations. I know. But it would require a profile fix to change that name so I just went with it.)